File: 235.lowest-common-ancestor-of-a-binary-search-tree.rb
#
# @lc app=leetcode id=235 lang=ruby
#
# [235] Lowest Common Ancestor of a Binary Search Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {TreeNode} p
# @param {TreeNode} q
# @return {TreeNode}
def lowest_common_ancestor(root, p, q)
  current_node = root

  while current_node
    if current_node.val < p.val && current_node.val < q.val
      current_node = current_node.right
    elsif p.val < current_node.val && q.val < current_node.val
      current_node = current_node.left
    else
      return current_node
    end
  end
end
# @lc code=end

File: 235.lowest-common-ancestor-of-a-binary-search-tree.go
/*
 * @lc app=leetcode id=235 lang=golang
 *
 * [235] Lowest Common Ancestor of a Binary Search Tree
 */

package main

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val   int
 *     Left  *TreeNode
 *     Right *TreeNode
 * }
 */

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	if p.Val < root.Val && q.Val < root.Val {
		return lowestCommonAncestor(root.Left, p, q)
	} else if p.Val > root.Val && q.Val > root.Val {
		return lowestCommonAncestor(root.Right, p, q)
	}

	return root
}

// func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
// 	if root.Val == p.Val {
// 		return p
// 	}
// 	if root.Val == q.Val {
// 		return q
// 	}

// 	if root.Val < p.Val && root.Val > q.Val {
// 		return root
// 	}
// 	if root.Val > p.Val && root.Val < q.Val {
// 		return root
// 	}

// 	if root.Val < p.Val {
// 		return lowestCommonAncestor(root.Right, p, q)
// 	} else {
// 		return lowestCommonAncestor(root.Left, p, q)
// 	}
// }

// @lc code=end
